### PR TITLE
feat(player): add loop method

### PR DESCRIPTION
**Pull request by Anto2441**
## Description

- Add loop method to the player (repeat the song)

Capture d’écran 2024-05-27 à 15 42 14

### CODE REVIEW

File: src/player/features/loop/ui/LoopButton.tsx [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    setIsRepeating((prevState) => {
+      const newRepeatingState = !prevState;
+      loopFeature.execute(newRepeatingState);
+
+      return newRepeatingState;
+    });
+  };
+
+  return (
+    <IconButton tooltip="Repeat" size="small" onClick={handleClick}>
```
Comment (carlo-ferreira) [inline src/player/features/loop/ui/LoopButton.tsx]: Not sure we need this line `const [isRepeating, setIsRepeating] = React.useState(false);`;

Maybe we can just execute the use case. Then read from `player.isLooping` directly?

File: src/player/features/loop/ui/LoopButton.tsx [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+      <PiRepeat
+        color={
+          isRepeating ? "var(--color-bright-accent)" : "var(--color-subdued)"
```
Comment (carlo-ferreira) [inline src/player/features/loop/ui/LoopButton.tsx]: And here we would have: `player.isLooping ? ... : ...`

By the way, do we have to use css custom properties here? Using tailwind classes doesn't work?

**Review by carlo-ferreira — APPROVED** [review]
Looks great 🏆

File: src/player/features/loop/ui/LoopButton.tsx (hunk 2 above)
Comment (Anto2441) [reply src/player/features/loop/ui/LoopButton.tsx]: Corrected !

File: src/player/features/loop/ui/LoopButton.tsx (hunk 1 above)
Comment (Anto2441) [reply src/player/features/loop/ui/LoopButton.tsx]: Yes good point, I use `this.isLooping = !this.isLooping` directly in the loop function in the `HowlerAudioManager`

Comment (carlo-ferreira) [reply src/player/features/loop/ui/LoopButton.tsx]: Nice 👌. No more setState, we let MobX doing his magic 🪄

File: src/player/lib/audio-manager/HowlerAudioManager.ts [hunk 3]
```diff
@@ -20,6 +22,12 @@ export class HowlerAudioManager implements IAudioManager {
     makeAutoObservable(this);
   }
 
+  loop(repeat: boolean) {
+    this.isLooping = !this.isLooping;
+
+    this.howl?.loop(repeat);
+  }
+
```
Comment (carlo-ferreira) [inline src/player/lib/audio-manager/HowlerAudioManager.ts]: Sorry, I haven't see this code.

We run the risk to be desynchronized between our HowlerAudioManager and Howl here.

We should either

```ts
loop(repeat: boolean) {
    this.isLooping = repeat;

    this.howl?.loop(repeat);
  }
```

or 

```ts
loop() {
    this.isLooping = !this.isLooping;

    this.howl?.loop(!this.isLooping);
}
```

See? Both our AudioManager and Howler will be updated with the **same state** no matter what the **initial state** is.

Comment (Anto2441) [reply src/player/lib/audio-manager/HowlerAudioManager.ts]: Yes, good point, thank you! 

Corrected.

**Pull request merged**